src/main/scala/io/druid/indexer/spark/SparkDruidIndexer.scala

```diff
@@ -636,24 +641,19 @@ class DateBucketAndHashPartitioner(gran: Granularity, partMap: Map[(Long, Long),
           throw new IAE(s"Unknown value type ${x.getClass} : [$x]")
       }
       val dateBucket = gran.truncate(new DateTime(k)).getMillis
-      val modSize = maxTimePerBucket.get(dateBucket.toLong)
-      if (modSize.isEmpty) {
-        // Lazy ISE creation
-        throw new ISE(s"bad date bucket [${dateBucket.toLong}]. " +
-          s"available: ${maxTimePerBucket.keySet}")
+      val shardLookup = shardLookups.get(dateBucket)
+      if (shardLookup.isEmpty) {
+        throw new IAE(s"Bad date bucket $dateBucket")
       }
-      val m = modSize.get
-      if (m <= 0) {
-        throw new ISE(s"Illegal mod [$m]")
+      val shardSpec = Option(shardLookup.get.getShardSpec(dateBucket, new MapBasedInputRow(dateBucket, if (dims == null) map.keys.toList.asJava else dims.toList.asJava, map.asJava)))
+      if (shardSpec.isEmpty) {
+        throw new ISE(s"Could not find date bucket for $dateBucket")
       }
-      val hash = if(dims != null) {
-        Math.abs(map.filterKeys(dims.contains).hashCode().toLong) % m.toLong
-      } else {
-        Math.abs(map.hashCode().toLong) % m.toLong
-      }
-      val partOpt = partMap.get((dateBucket.toLong, hash.toLong))
+      val timePartNum = shardSpec.get.getPartitionNum
+
+      val partOpt = partMap.get((dateBucket.toLong, timePartNum))
```

Comment:
as per prior comment, any time I did the getOrElse throw new exception technique it always made the exception regardless of if the throw branch was taken.